### ISSUE TITLE
[oauth2] Can't get token using authorization_code

### ISSUE THREAD

**Issue by gs11** [gravitee-io/graviteeio-access-management#435]
I'm trying to get the authorization code flow working but can't get a token using an authorization_code
I successfully get an authorization code but when requesting the token I only get a 401 Unauthorized.

authorize request is 
`GET http://localhost:8092/k2/oauth/authorize?client_id=gui&redirect_uri=http://localhost:1234&response_type=code&scope=test&state=dummy`
token request is
```
POST http://localhost:8092/k2/oauth/token
    grant_type:"authorization_code"
    code:"<auth code>"
    client_id:"gui"
    redirect_uri:"http://localhost:1234"
    scope:"test"
```

The [API docs](https://docs.gravitee.io/am/2.x/oauth2/index.html#tag/OAuth-2.0%2Fpaths%2F~1token%2Fpost) state that username, password & refresh_token is mandatory but they shouldn't be in an SPA authorization code scenario?

Token requests work just fine using client_credentials grant type

I've created:
- An inline identity provider with my user that logs in during authorization
- A client "gui" with default settings and linked to the above inline identity provider. This was successfully used for the client_credentials token request
- A scope "test"

Am I doing something fundamentally wrong here for the authorization code grant type and can I activate some debug logging to see why I'm getting the 401?

**Comment by tcompiegne** [comment]
Hi @gs11 , 

Token endpoint is protected with a Basic Authentication

```
client_auth
Base64(clientId:clientSecret)
```

Did you set it right ? 

Regards

**Comment by gs11** [comment]
But it shouldn't be needed as I can't/don't want to store any credentials in the web application @tcompiegne

**Comment by tcompiegne** [comment]
If you are using Single Page Application (SPA) you should use `Implicit Grant` (response_type=code) to directly retrieve an Access Token and not expose your client credential.

**Comment by gs11** [comment]
Thanks, but the authorization code grant type should also work without involving any credentials with the token endpoint. It is the case with all other OAuth providers I've used.

For now I can stick to the implicit grant though.

**Comment by tcompiegne** [comment]
You are right, currently Gravitee.io assumes that the client is "confidential" and requires a client authentication.

But you can create an issue to handle "Public Client", what do you think ? 

Regards

**Comment by gs11** [comment]
Absolutely, but maybe it could be handled in this issue?

**Comment by tcompiegne** [comment]
Yes, I rename the issue
